modifications for round

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

closes #12056

#### Brief description of what is fixed or changed

* The round method has been modified to round to even by default
in keeping with Python 3 semantics

* The round method no longer casts args to float so precision
is no longer lost.

* To get Python 3 behavior in Python 2, import round from sympy.core.compatibility


#### Other comments

Python and SymPy use different methods of rounding so rounding to the right of the decimal is WYSIWIG in SymPy but not in Python:
```python
>>> round(12.345,2)
12.35
>>> S(12.345).round(2)
12.34
```
This is because Python uses the full binary representation while SymPy rounds an integer based on the precision of the number, e.g. if a number `n` is stored as 0.3499999 with 2 digits of precision, SymPy would use 35 for the rounding since this is the closest integer to `n*10**2`.

cf [discussion here](https://groups.google.com/forum/#!topic/sympy/CgkxNTT85aI)

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
  - round added to compatibility
  - under Python 3, Number.round == round(Number)
  - Number.round rounds to even on tie and does so in cases that Python misses, e.g. round(12.345,2) -> 12.35 in Python but 12.34 in SymPy whereas round(4.5) -> 4 in both.
  - Integer input is now returned as Integer

<!-- END RELEASE NOTES -->

smichr commented:
@asmeurer, do you want this in 1.4 or should it wait?

asmeurer commented:
It's too late for 1.4 unless it is an absolutely critical bug fix. I plan to do the final release tomorrow or Wednesday.

smichr commented:
> absolutely critical bug fix

Nope -- just creating a compatibilty version of round and changing default rounding semantics

Sc0rpi0n101 commented:
Some of the doctests are failing 
for eg, in pi.round()
it's Expected: 3
Got: 3. not 3.0 or 3
It's with the single number results only. All others are like 2.0*I or 6.0*3.0*I(still failing as int was expected)
is this a bug with Travis?

smichr commented:
> Some of the doctests are failing

Thanks -- updated accordingly:
```
sympy\core\expr.py[33] .................................                   [OK]
```

smichr commented:
It would be good to have another set of eyes check the changes. 

Does `how` work as a rounding method keyword? An alternative would be `rnd` but that looks like "random" to me...but that is the variables used in `evalf.py`

asmeurer commented on sympy/core/expr.py:
```diff
@@ -3406,12 +3419,23 @@ def round(self, p=0):
+            if 2*(i - self*mag) == 1 and i%2:
+                # round toward even
+                rv = Rational(i - 1, mag)
+        elif how != 'up':
```
"up" is ambiguous. Does this mean towards +infinity or away from 0?

smichr replied:
I vote for not supporting custom rounding. That way the round function and method are synonymous in behavior.

smichr replied:
But to answer the question, it rounds away from 0.

asmeurer replied:
That's fine too. There are many rounding methods which could be added, but they don't need to be done in this PR.

asmeurer commented on sympy/core/expr.py:
```diff
@@ -3325,50 +3325,63 @@ def invert(self, g, *gens, **args):
         return invert(self, g, *gens, **args)
 
-    def round(self, p=0):
+    def round(self, p=0, how='even'):
```
I would use a more descriptive name for the keyword, like `rounding_type`.

asmeurer commented:
Some tests that `round(Float)` matches `round(float)` in Python 3 would be good.

smichr commented:
I have to look at the irregular output I am getting.

smichr commented:
I'm not sure what to do about rounding Integers. What should the following give and why?
```
>>> round(123)  # matches Python single-argument rules
123
>>> [round(S(123), i) for i in (-1,0,1)]  # preserves precision information
[1.2e+2, 123., 123.0]
```

Sc0rpi0n101 commented:
Shouldn't round(123,0) be 123, an int 
instead of 123. , a float with 0 decimal places?

smichr committed:
round changes

- don't recast arg to float
- Integer in Integer out
- single arg call (n=None) gives Integer out

smichr commented:
See OP for how SymPy and Python differ in rounding algorithms.

@asmeurer , comparisons of types are now tested.

smichr commented:
> Shouldn't round(123,0) be 123, an int

In Python 3, yes. This is now the result given by SymPy in this branch.

smichr commented:
wip to make sure all SymPy numbers round to even on tie with given precision

smichr commented:
wip to make sure all SymPy numbers round to even on tie with given precision

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1676,13 +1676,28 @@ def test_round():
 
     assert (Float(.3, 3) + 2*pi).round() == 7
     assert (Float(.3, 3) + 2*pi*100).round() == 629
-    assert (Float(.03, 3) + 2*pi/100).round(5) == 0.09283
```
the output is limited by the precision of 3

smichr committed:
round result is always float in Py2

smichr commented:
I feel like the test that were in place and those that I have added are a good test of this algorithm. I am also happy that we have round-to-even rounding working for Floats. The simplified code here belies the difficulty of getting this right. If anyone else wants to give this a check that would be nice, otherwise I will commit this post-Easter.

oscarbenjamin commented:
+1

Should a new issue be opened with the dropping Python 2 label noting to follow up on this later?

smichr commented:
cf #16686

cbm755 commented:
I'd like to (eventually) see Float rounding done by mpmath (https://github.com/fredrik-johansson/mpmath/issues/455)

Then we could worry separately about how Rationals, irrational Numbers, etc are dealt with, without the code for Floats being interspersed.  But that could wait until Python 2 is dropped, so I'm ok in principle with some of these changes.

- - - -

But the round-to-even bit is suspicious for Floats.  I don't believe the "augment precision, add 5 to previously least significant digit" logic.

Here is a float, converted to a Float, that is less than 575/1000 but none-the-less rounds (incorrectly) to 0.58:
```
>>> a = Float(0.575 - 6e-16)
>>> a < Rational(575, 1000)
True
>>> round(a, 2)
0.58
```

smichr committed:
add rounding notes; use tenths digit for rounding

smichr commented:
@cbm77, please see inline notes to the algorithm which is now recast more explicitly in terms of known digits as an integer and the rounding digit in the 10ths place.
